lib/nimble_parsec/compiler.ex

```diff
@@ -683,6 +683,11 @@ defmodule NimbleParsec.Compiler do
     end
   end
 
+  defp bound_combinator(:eof, line, offset, counter) do
+    guard = quote(do: byte_size(rest) == 0)
```

Comment:
Is it preferable to check `rest == ""` or should we prefer an unbounded combinator that has `<<>>` as first argument in function head. This is a simpler implementation than using unbounded combinator as we know the length.

Reply:
Unfortunately, I think both `when byte_size(_) == 0` and `when _ == ""` don't work well with the binary patterns optimiser. The ideal version would pattern match on `<<>>` in a clause.

Reply:
We still want to keep `eof` as a bound combinator so that it can be used for the single pattern cases - we know its length. This can be achieved by pushing `rest :: binary` into the `inputs` for bound combinators. Thoughts?

Reply:
I'm not sure I understood your suggestion.

Reply:
If we want to use bound combinator for this change, and match on `<<>>` (or potentially `<<rest::binary-size(0)>>` to maintain the variable), the bound combinators need to control the full binary pattern. Right now the `inputs` (var name used throughout module) only has the patterns that produce outputs, and no control over `rest`.

Reply:
Ah, makes sense. I'm not sure we can make an efficient eof combinator without controlling the size of `rest` in that case.